tests/runtime/test_pinned_buffers.cpp

```diff
@@ -34,21 +34,42 @@
 
 #define BUF_SIZE 16
 
+// A kernel that gets the device-seen address of the buffer.
 static char GetAddrSourceCode[] = R"raw(
 
-  __kernel void get_addr (__global int *pinned_buffer,
+  __kernel void get_addr (__global int *buffer,
                           __global ulong* addr) {
     for (int i = 0; i < BUF_SIZE; ++i)
-      pinned_buffer[i] += 1;
-    *addr = (ulong)pinned_buffer;
+      buffer[i] += 1;
+    *addr = (ulong)buffer;
   }
 )raw";
 
-void *getDeviceAddressFromHost(cl::Buffer &Buf) {
-  cl_mem_pinning Pinning;
+// A kernel that accesses another buffer indirectly.
+static char IndirectAccess[] = R"raw(
+
+  __kernel void indirect_access (__global long* in_addr,
+                                 __global int* out) {
+    *out = **(int**)in_addr;
```

Comment:
For runtimes not supporting generic pointers this needs to be `*out = **(int __global* __global*)in_addr;` or whatever style you prefer to specify the address space (I just put them right before the `*` because that feels more natural to me).